core/src/main/scala/org/apache/spark/sql/catalyst/expressions/ParamLiteral.scala

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (c) 2016 SnappyData, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you
+ * may not use this file except in compliance with the License. You
+ * may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ */
+
+package org.apache.spark.sql.catalyst.expressions
+
+import java.util.Objects
+
+import com.esotericsoftware.kryo.io.{Input, Output}
+import com.esotericsoftware.kryo.{Kryo, KryoSerializable}
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.codegen.{CodegenContext, ExprCode}
+import org.apache.spark.sql.types._
+
+case class ParamLiteral(l: Literal, pos: Int) extends LeafExpression {
```

Comment:
I think this should inherit from Literal instaed of LeafExpression so that rest of the spark code doing case match on Literal won't stop matching. One such sideeffect I am observing with this change is filter push down. I suppose we don't have enough test validating filters are practically pushed down.

effected code snippet
```scala
DataSourceStrategy.scala
protected[sql] def translateFilter(predicate: Expression): Option[Filter] = {
    predicate match {
      case expressions.EqualTo(a: Attribute, Literal(v, t)) =>
        Some(sources.EqualTo(a.name, convertToScala(v, t)))
      case expressions.EqualTo(Literal(v, t), a: Attribute) =>
        Some(sources.EqualTo(a.name, convertToScala(v, t)))
 . . . 
```
used in StoreDataSourceStragy.scala
```scala
  private def pruneFilterProjectRaw(
      relation: LogicalRelation,
      projects: Seq[NamedExpression],
      filterPredicates: Seq[Expression],
      numBuckets: Int,
      partitionColumns: Seq[String],
      scanBuilder: (Seq[Attribute], Seq[Expression], Seq[Filter]) =>
          (RDD[Any], Seq[RDD[InternalRow]])) = {
 . . . 
    val (unhandledPredicates, pushedFilters) =
      selectFilters(relation.relation, candidatePredicates)
 . . . 
```
Above code now no more matches the constant in the where clause for a query like below
```sql
    snc.sql("select * from orders where o_orderkey = 1 ").collect().foreach(println)
    snc.sql("select * from orders where o_orderkey = 2 ").collect().foreach(println)
    snc.sql("select * from orders where o_orderkey = 3 ").collect().foreach(println)
```